libsolidity/lsp/SemanticTokensBuilder.cpp

```diff
@@ -0,0 +1,289 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libsolidity/lsp/SemanticTokensBuilder.h>
+
+#include <liblangutil/CharStream.h>
+#include <liblangutil/SourceLocation.h>
+
+#include <fmt/format.h>
+
+using namespace std;
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+#if 0 // !defined(NDEBUG)
+#define DPRINT(x) do { std::cerr << (x) << std::endl; } while (0)
+#else
+#define DPRINT(x) do {} while (false)
+#endif
+
+namespace solidity::lsp
+{
+
+namespace
+{
+
+optional<SemanticTokenType> semanticTokenTypeForType(frontend::Type const* _type)
+{
+	if (!_type)
+		return nullopt;
+
+	switch (_type->category())
+	{
+	case frontend::Type::Category::Address: return SemanticTokenType::Class;
+	case frontend::Type::Category::Bool: return SemanticTokenType::Number;
+	case frontend::Type::Category::Enum: return SemanticTokenType::Enum;
+	case frontend::Type::Category::Function: return SemanticTokenType::Function;
+	case frontend::Type::Category::Integer: return SemanticTokenType::Number;
+	case frontend::Type::Category::RationalNumber: return SemanticTokenType::Number;
+	case frontend::Type::Category::StringLiteral: return SemanticTokenType::String;
+	case frontend::Type::Category::Struct: return SemanticTokenType::Struct;
+	case frontend::Type::Category::Contract: return SemanticTokenType::Class;
+	default:
```

Comment:
If you keep the default case here, we will not get a compiler erorr when we add a new type category.